feat: person-overrides writes with Mapping to Persons

tomasfarias opened the pull request:
## Problem

closes https://github.com/PostHog/posthog/issues/13776
So, this PR attempts to incorporate the changes from https://github.com/PostHog/posthog/pull/14368 and https://github.com/PostHog/posthog/pull/14397, but on top of https://github.com/PostHog/posthog/pull/14477.

Yikes.

For some commits, I was able to cherry-pick them as they are. Some others, had too many conflicts with recent changes so they had to be manually re-done.

The solution in those two PRs attempted to use a CHECK constraint to enforce consistency while merging overrides. This was problematic as it required a FK on very hot tables (`Person` and `Team`), which made it very hard to the deploy.

This solution uses an exclusion constraint instead, but since exclusion constraints on `uuid[]` are not supported, we created a helper model in #14477 that maps a serial int id to person UUIDs. Exclusion constraints on `int[]` are possible with two extensions.

<!-- Who are we building for, what are their needs, why is this important? -->

## Changes

On top of the new helper model, this PR includes the merging logic. This logic has been modified from #14368 to utilize the `PersonOverridesMapping` model. This requires two more queries to insert into the helper model as we merge.

Moreover, this also includes a fix for the latest migration as when dropping a field a check constraint was dropped. So it has to be re-added.

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

## How did you test this code?

Unit tests were mostly kept the same.
<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

Commit:
wip: add test for concurrent updates to posthog_personoverride table

Commit:
refactor

github-actions[bot] committed:
Update snapshots

tomasfarias committed:
fix(isort): Correctly sort imports

tomasfarias committed:
fix(ee-cohort-test): Delete person after creating it

tiina303 committed:
fix: person constraint in person overrides table (#14319)

tiina303 committed:
nits

tiina303 committed:
fix migration tests

tomasfarias committed:
chore(migration): Bump migration number to 0302

github-actions[bot] committed:
Update snapshots

tiina303 committed:
feat: person-overrides writes

tomasfarias committed:
fix(tests): Pass poEEmbraceJoin to updatePersonState in test

tiina303 committed:
fix: poe final test failure

tomasfarias committed:
refactor(person-state): Make failed attempts a class variable to

This allows us to mock it during testing as some tests require
immediate failures.

tomasfarias committed:
feat(person-override): Add a helper model to indirectly reference person overrides

This allows us to use an exclusion constraint on the person overrides table instead of directly using a FK on posthog_person.

tomasfarias committed:
feat(person-override): Add a helper model to indirectly reference person overrides

This allows us to use an exclusion constraint on the person overrides table instead of directly using a FK on posthog_person.

tomasfarias committed:
test(person-overrides): Update tests to match new constraint

tomasfarias committed:
test(person-overrides): Add concurrent tests to person-overrides model

tomasfarias committed:
feat(person-merge): Update merge to use new helper table

tomasfarias committed:
fix(person-state): Format person-overrides message for ClickHouse

tomasfarias committed:
test(api): Add function to reload dictionary for person overrides

tomasfarias committed:
fix(postgres-utils): Apply some magic changes lost to time

tomasfarias committed:
fix(person-state): Join with helper table to return UUIDs

tomasfarias committed:
fix(person-state): Use single quotes for UUID queries

tomasfarias committed:
test(person-overrides): Skip test that doesn't work without a merge command

tomasfarias committed:
test(person-state): Add a very complicated query to get UUIDs in a test

tomasfarias committed:
test(person-state): Try waiting longer, this is flaky

tomasfarias committed:
fix(migrations): Remove unused migration

tomasfarias committed:
fix(person): Undo unneeded person model changes

tomasfarias committed:
fix(migration): Add drop extension query to reverse migration

tomasfarias committed:
fix(migration): Use correct table name

tiina303 commented on posthog/models/person/person.py:
```diff
@@ -121,10 +121,11 @@ class PersonOverride(models.Model):
     2. Check that old_person_id is different to override_person_id for every row.
-    3. Exclude rows that overlap across old_person_id and override_person_id (e.g. if
-        a row exists with old_person_id=123 then we would not allow a row with
+    3. Same person id cannot be used as an old_person_id and an override_person_id (per team)
```
these changes should be in https://github.com/PostHog/posthog/pull/14477

tomasfarias replied:
I've moved this to #14477 and rebased this PR.

tiina303 commented on posthog/models/test/test_person_override_model.py:
```diff
@@ -1,11 +1,15 @@
+import contextlib
```
these changes should be in https://github.com/PostHog/posthog/pull/14477

tomasfarias replied:
I've moved these changes to #14477 and rebased this PR.

tiina303 commented on posthog/management/commands/test/test_migrate_kafka_data.py:
```diff
@@ -1,347 +0,0 @@
-from unittest import mock
```
can we pull this out of this PR - maybe a separate pr on master if the tests still fail

tomasfarias replied:
The problem is tests do not fail. They run forever until GitHub actions times out. And it only happens when mixed in with these changes.

tomasfarias replied:
So, if we separate it then:
1. In this PR tests will run forever.
2. In the other PR everything will run fine.

tomasfarias replied:
Haven't had any luck in catching the bug :sweat:

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -295,6 +306,8 @@ export class PersonState {
         } catch (e) {
             console.error('handleIdentifyOrAlias failed', e, this.event)
+
+            throw e
```
iiuc: This would mean that if someone sent us an identify and we decided not to merge we would drop that event, we shouldn't

tomasfarias replied:
Why not? Nothing was merged.

tiina303 replied:
We should add a test for this: make sure the events gets ingested when merge was blocked (easiest to simulate with illegal ids) and also make sure properties get updated for the distinct_id user.

This is the current behaviour and I don't see why we should change it. We have quite a lot of blocked merges, it's easier to debug if we end up processing that event, also if you think of CDP all events (whenever possible) should be exported.

tomasfarias replied:
I'm working on adding a test for the case you mentioned.

tomasfarias replied:
I've dropped this `throw e`, and added a couple of test cases to verify events are being ingested on a failed `$merge_dangerously` and that properties are updated. Let me know if these look like you expected or I should do any updates.

Retrying PluginServer CI a few times due to flaky tests...

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -506,10 +526,168 @@ export class PersonState {
+
+         1. Add the persons involved to the helper table (2 of them)
+         2. Add an override from oldPerson to override person
+         3. Update any entries that have oldPerson as the override person to now also point to the new override person
+
+            TODO: how do we want to be updating oldest_event?
+
+         I'm thinking: we write it if it's not there, but don't update it on conflicts (alternative update to the older date)
+         In the transitive one the same don't update or update to older data (if that's not too complex to do)
+         it's an optimization anyway and squash job might be updating those values anyway and it's just a hint, so ignoring if complex seems better
+         in any case it's important that we use the oldest_event we stored in postgres in CH too.
```
```suggestion
         3. Update any entries that have oldPerson as the override person to now also point to the new override person. Note that we don't update `oldest_event`, because it's a heuristic (used to optimise squashing) tied to the old_person and nothing changed about the old_person who's events need to get squashed.

            TODO: how do we want to be updating oldest_event?

         I'm thinking: we write it if it's not there, but don't update it on conflicts (alternative update to the older date)
         In the transitive one the same don't update or update to older data (if that's not too complex to do)
         it's an optimization anyway and squash job might be updating those values anyway and it's just a hint, so ignoring if complex seems better
         in any case it's important that we use the oldest_event we stored in postgres in CH too.
```

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -506,10 +526,168 @@ export class PersonState {
+                    ${oldPersonId},
+                    ${overridePersonId},
+                    ${oldestEvent},
+                    1
```
why not 0?

tomasfarias replied:
This commit was cherry-picked from your PR here: https://github.com/PostHog/posthog/pull/14368/files#diff-c51e19d65898f9cd1b4aaad32e5163e7b44507026ce49301507a2cbd6fe5ccdbR572

Which originally came from @hazzadous PR? Not sure about this.

Regardless, my point is I have no idea why 1 here and not 0. So, I can just do that change.

tomasfarias replied:
Made it 0.

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -506,10 +526,168 @@ export class PersonState {
+                    ${oldestEvent},
+                    1
+                )
+                RETURNING version;
```
that's not necessary - it's the hardcoded value

tomasfarias replied:
Same answer as above, change come from your PR. Happy to get rid of this `RETURNING` clause.

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -506,10 +526,168 @@ export class PersonState {
+                    UPDATE
+                        posthog_personoverride
+                    SET
+                        override_person_id = ${overridePersonId}, version = version + 1
```
This should be `version = COALESCE(version, 0)::numeric + 1` like we do for distinct_id moves. This is important for concurrent merges that might end up with two overrides with the same version - this would be a problem as both Kafka messages end up going to CH and the latter would randomly pick one of them to use (not what we want).

tomasfarias replied:
I believe it's impossible for concurrent merges to run this query simultaneously as they both would lock the table, so only one would be allowed to proceed. The one that wasn't would just read the new updated value that the previous query set.

That being said, I don't see anything wrong with adding the `COALESCE` call (although it did made me question why we even allow `version` to be `null`).

tiina303 replied:
hmm... yeah I'm not a postgres expert, but consistency with pdi is probably good anyway

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -506,10 +526,168 @@ export class PersonState {
+                )
+                SELECT * FROM insert_id
+                UNION ALL
+                -- ON CONFLICT nothing is returned, so we get the id here.
```
we have a test for this case, right?

tomasfarias replied:
The `handles a chain of overrides being applied concurrently` test has `$identify` calls that would conflict here as both of them are mapping to the same person.

tiina303 replied:
well I'm just waiting on this PR to land so I can make the changes to allow merges, don't want to create conflicts for you :)

tomasfarias replied:
I am very confused by your reply. We handle this conflict in the query, it is expected to happen during normal merges.

tomasfarias replied:
Of course, the merge may fail other constraints down the line, but that's not what this comment references.

tiina303 commented on plugin-server/tests/worker/ingestion/person-state.test.ts:
```diff
@@ -2057,3 +2097,488 @@ describe('PersonState.update()', () => {
+    beforeEach(async () => {
+        const organizationId = await createOrganization(hub.db.postgres)
+        teamId = await createTeam(hub.db.postgres, organizationId)
+        // jest.useFakeTimers({ advanceTimers: 50 })
```
^

tomasfarias replied:
No idea what "^" means. Delete this comment?

tiina303 replied:
"^" here meant you probably don't want to merge like this, can you figure out how this test should be fixed or if the comment should be removed or I don't know exactly what needs to happen, but we don't want to merge commented out code generally - it was here because it was WIP initially I believe

tomasfarias replied:
I'm fine with deleting the comment, test is passing so I don't think there's anything to fix.

tiina303 commented on plugin-server/tests/worker/ingestion/person-state.test.ts:
```diff
@@ -2057,3 +2097,488 @@ describe('PersonState.update()', () => {
 })
+
+describe('person id overrides', () => {
+    // For some reason these tests failed if I ran them with a hub shared
+    // with other tests, so I'm creating a new hub for each test.
+    let hub: Hub
```
can you retry to see if that's still the case?

tiina303 commented on plugin-server/tests/worker/ingestion/person-state.test.ts:
```diff
@@ -2057,3 +2097,488 @@ describe('PersonState.update()', () => {
+            .sort() as [string, string][]
+    }
+
+    it('postgres overrides added on merge', async () => {
```
Looks like there are no tests for CH side - can we add some

tomasfarias replied:
What do you mean by "CH side"? Ingestion tests query ClickHouse.

tiina303 replied:
yes like the other tests earlier in this file - they verify that CH state is right too.

tomasfarias replied:
Looks like it works in CI.

tiina303 commented on plugin-server/tests/worker/ingestion/person-state.test.ts:
```diff
@@ -2057,3 +2097,488 @@ describe('PersonState.update()', () => {
+                    uuid: person.uuid,
+                    properties: person.properties,
+                    created_at: person.created_at,
+                    // TODO: fix this. It seems that we update some person
```
^

tomasfarias replied:
Would you mind expanding on "^"?

This commit was also not added by me, but cherry picked from other PR (either yours or @hazzadous's).

So, if your "^" means "fix this what this comment says" then I'm going to need more context on that as I didn't write the comment myself.

tiina303 replied:
I don't know the context, but can look into it if needed, I'm not even sure it's true 🤷

hazzadous replied:
If you uncomment the lines is_identified and version, you'll end up with:

```
    +   Object {
    -     "is_identified": false,
    +     "is_identified": true,
...
    -     "version": "0",
    +     "version": "1",
        },
```

So if we for some reason fail to completely process the event, on retry we will see `is_identified === true` and thus we will not properly handle the retry as there is code that depends on the comparison. It's like that already, nothing to do with this PR so fine to leave.

The main thing to test is that running it again gets us to the right state which I think we do in this case. However, I _think_ there are cases where it will not be correct, e.g. if one of the persons was already identified, then the second becomes identified, then we can merge them afterwards on second run.

tiina303 replied:
aah I get it now, this is deterministic, so we can test this properly. We always mark the distinct_id person as `is_identified` even if merge failed. This helps make sure that anything that's a user_id won't be merged into anything else via `$identify` nor `$create_alias`.

So the test can be: 
```
distinct_id == 'first' person: is_identified: true, version: 1
distinct_id == 'second' person: is_identified: false, version: 0
```

Potentially easiest to change the code by adding a person property with the distinct_id values 'first' and 'second'

tomasfarias committed:
refactor(person-overrides): Make team a regular bigint field without FK

tiina303 reviewed (commented):
`which made it very hard to the deploy.` => I'd add that going back to the drawing board we realized there's a more explicit way we can validate correctness.

tomasfarias committed:
refactor(person-overrides): Rename Helper model to Mapping

tomasfarias committed:
chore: Clean-up artifacts from rebase on model branch

tomasfarias commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -62,14 +67,17 @@ export class PersonState {
-        uuid: UUIDT | undefined = undefined
+        poEEmbraceJoin: boolean,
+        uuid: UUIDT | undefined = undefined,
+        mergeAttemps: number = MAX_FAILED_PERSON_MERGE_ATTEMPTS
```
Yeah, having the `max` there is clearer.

tomasfarias commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -6,16 +6,19 @@ import { ProducerRecord } from 'kafkajs'
 import { captureIngestionWarning } from './utils'
 
-const MAX_FAILED_PERSON_MERGE_ATTEMPTS = 3
+const MAX_FAILED_PERSON_MERGE_ATTEMPTS = process.env.MAX_FAILED_PERSON_MERGE_ATTEMPTS
+    ? Number(process.env.MAX_FAILED_PERSON_MERGE_ATTEMPTS)
+    : 3
```
Cool, I'll get rid of the env code.

tomasfarias committed:
refactor(person-state): Rename mergeAttempts and don't read from ENV

tomasfarias committed:
chore: Better clarify oldest_event usage in comment

Co-authored-by: Tiina Turban <tiina303@gmail.com>

tomasfarias committed:
refactor(person-state): Update version in queries

tomasfarias committed:
fix(person-state): Use new mapping model instead of helper

tomasfarias committed:
fix(migrations): Re-add constraint deleted on field drop

tomasfarias committed:
revert: Re-enable test that was skipped

Test was originally skipped due to missing a $merge command, but now
$merge_dangerously is available, so the test should pass.

tomasfarias committed:
test: Attempt to refresh dictionary in test

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

tomasfarias committed:
test: Expect dictionaries to be refreshed before resuming test

tomasfarias committed:
fix(test): Use alias property for $merge_dangerously in test

tomasfarias committed:
fix(test): Expect any string like in all other tests

tiina303 commented on plugin-server/functional_tests/ingestion.test.ts:
```diff
@@ -492,16 +501,22 @@ testIfPoEEmbraceJoinEnabled(`chained merge results in all events resolving to th
 })
 
+// This test only works when poEEmbraceTheJoin is enabled.
+// For this test to pass we must write a $merge command that works when both persons are identified.
```
Remove the comment line

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -504,10 +522,165 @@ export class PersonState {
+                )
+                SELECT * FROM insert_id
+                UNION ALL
+                -- ON CONFLICT nothing is returned, so we get the id here.
```
do we want this comment to be within the query - I'd move it outside before

tomasfarias replied:
I would need to update the comment as "here" makes sense in the query as the `SELECT` it's literally right after the comment. Otherwise, I'm neutral about this.

tiina303 commented on plugin-server/tests/worker/ingestion/person-state.test.ts:
```diff
@@ -2117,3 +2157,488 @@ describe('PersonState.update()', () => {
+        await closeHub()
+    })
+
+    async function updatePersonStateFromEvent(event: Partial<PluginEvent>, ts = '', mergeAttempts = 3) {
```
lets please default to 1, so it's easier to debug and fails faster

tomasfarias committed:
test: Fix order of events to determine merged persons

tomasfarias committed:
chore: Update comment regarding overrides mapping query

tomasfarias committed:
chore: Remove out of date comment

tomasfarias committed:
test: Set number of retries to 0 when updating persons

tomasfarias committed:
test: Delete comment

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -504,10 +522,165 @@ export class PersonState {
+         2. Add an override from oldPerson to override person
+         3. Update any entries that have oldPerson as the override person to now also point to the new override person. Note that we don't update `oldest_event`, because it's a heuristic (used to optimise squashing) tied to the old_person and nothing changed about the old_person who's events need to get squashed.
+
+            TODO: how do we want to be updating oldest_event?
```
https://github.com/PostHog/posthog/pull/14513#discussion_r1126292605

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -504,10 +522,167 @@ export class PersonState {
+                        version,
+                    oldest_event
+                )
+                -- The follow-up JOIN is required as ClickHouse requires UUIDs, so we need to fetch the UUIDs
```
maybe this one also before the query

tomasfarias replied:
Same comment as before, I'm neutral about this.

tomasfarias committed:
test: Assert clickhouse state after identify

tomasfarias committed:
chore: Move comment

tomasfarias committed:
chore: Delete TODO regarding oldest_event updates

tomasfarias committed:
test: Assert events are still processed if merge fails

tomasfarias committed:
test: Assert properties are processed if merge fails

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -504,10 +520,165 @@ export class PersonState {
+         1. Add the persons involved to the helper table (2 of them)
+         2. Add an override from oldPerson to override person
+         3. Update any entries that have oldPerson as the override person to now also point to the new override person. Note that we don't update `oldest_event`, because it's a heuristic (used to optimise squashing) tied to the old_person and nothing changed about the old_person who's events need to get squashed.
+
+         I'm thinking: we write it if it's not there, but don't update it on conflicts (alternative update to the older date)
+         In the transitive one the same don't update or update to older data (if that's not too complex to do)
+         it's an optimization anyway and squash job might be updating those values anyway and it's just a hint, so ignoring if complex seems better
+         in any case it's important that we use the oldest_event we stored in postgres in CH too.
```
```suggestion
```

tomasfarias committed:
test: Don't expect to throw when failing merges

hazzadous commented on plugin-server/tests/worker/ingestion/person-state.test.ts:
```diff
@@ -2117,3 +2167,516 @@ describe('PersonState.update()', () => {
+
+    beforeAll(async () => {
+        ;[hub, closeHub] = await createHub({})
+        await hub.db.clickhouseQuery('SYSTEM STOP MERGES')
```
What's the reason for this and should we also start merges in `afterAll`?

hazzadous reviewed (approved):
Had a quick skim, looks great.

I really enjoyed the the error cases tests on this one, but pride comes before a fall so I'm prepared 😆

tomasfarias committed:
fix: Indent RETURNING query clause

Co-authored-by: Harry Waye <harry@posthog.com>

tomasfarias committed:
chore: Remove out of date comment

Co-authored-by: Tiina Turban <tiina303@gmail.com>

tomasfarias committed:
test: Resume merges after test suite

tomasfarias commented:
Builds are broken due to a dependency issue, I'm rebasing on `master` to take care of it.

tiina303 reviewed (approved):
Could you just make the 1 change to run 
plugin-server/tests/worker/ingestion/event-pipeline/processPersonsStep.test.ts 
in both modes, other stuff we can do in follow-ups

tomasfarias committed:
test: Run processPersonStep tests for both poEEmbraceJoin modes
